# <div style='padding:25px;background-color:maroon;color:white;border-radius:4px;font-size:100%;text-align: center'>Banking Data Analysis<br></div>

## <span style="color:Aqua;"> Objective of the Project:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; The objective of this project is to leverage advanced data analytics
techniques, including classification, regression, and clustering, to extract valuable
insights and enhance decision-making processes within the banking sector. The
project aims to address various aspects of banking operations, such as customer
segmentation, credit risk assessment, and performance prediction.
### <p style="color:Aqua;"> Key Components:</p>

<p style="color:Tomato;font-size: 110%"> <b> 1. Customer Segmentation (Clustering):</b> </p>

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Utilize clustering algorithms to group customers based on their banking
behaviors, transaction histories, and demographics.

<p style="color:Tomato;font-size: 110%"> <b> 2. Credit Risk Assessment (Classification):</b> </p>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Develop a robust credit scoring system using classification algorithms to
assess the creditworthiness of loan applicants.<br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Utilize historical data to train the model and predict the likelihood of default
or late payments.<br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Enhance risk management by integrating machine learning models into the
credit approval process.

<p style="color:Tomato;font-size: 110%"> <b> 3. Performance Prediction (Regression):</b> </p>

<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Build regression models to predict key performance indicators (KPIs) for the
banking institution, such as asset growth, revenue, and profitability.<br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Analyze the impact of various factors, such as economic indicators and market
trends, on the bank's performance.<br>
<span style="color: Chartreuse;">   &#9784; &nbsp;</span> Provide actionable insights to optimize resource allocation, investment
strategies, and overall business performance.


## <span style="color:Aqua;">Importing libraries from Python</span>

In [361]:
import pandas as pd 
import os
import numpy as np
from IPython.display import display, HTML

pd.options.display.max_columns = 50
pd.set_option("display.precision", 4)
pd.set_option('display.float_format', '{:.4f}'.format)


## <span style="color:Aqua;">Data Exploration and Preprocessing:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> In this section we are gonna explore features of each variables (data columns) and address any issues with the data which may affect our Machine Learning model (Predicting app).



### <span style="color:Tomato;">Reading Dataset from CSV:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Reading data from the excel to pandas (Python's Data Wrangler).

In [362]:
df = pd.read_csv('train.csv')

### <span style="color:Tomato;"> Understanding the dataset:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output below we see that we have 1 Lakh rows and 28 columns. Next output is 5 sample data points generated randomly from the dataset.

In [363]:
display(HTML(f"<p style='color: orange; font-weight: bold;'>{df.shape}</p>"))
df.sample(5)

ID  Customer_ID  Month           Name     Age            SSN  \
83285  130559        31306      6      ita Bosex 19.0000  12494059.0000   
80389  126215        25003      6      Scuffhamf 33.0000 670224819.0000   
93445  145799        40914      6          Wakiy 41.0000  26142634.0000   
22214   38952        29315      7         Kareno 24.0000  78848943.0000   
60791   96817        42522      8  Marton Dunaig 23.0000  52733081.0000   

      Occupation  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
83285  Architect     58832.1900              4827.6825             8.0000   
80389     Doctor     19124.4000              1859.7000             5.0000   
93445   Musician     35644.7200              2987.3933             1.0000   
22214     Writer     53370.4200              4618.5350             6.0000   
60791    Teacher     20538.7700              1957.5642             8.0000   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  \
83285           3.0000        10.0000       3.0000   
80389           8.0000         7.0000       1.0000   
93445           4.0000         3.0000       1.0000   
22214           7.0000         7.0000       3.0000   
60791          10.0000        23.0000       6.0000   

                                            Type_of_Loan  Delay_from_due_date  \
83285  Student Loan, Debt Consolidation Loan, and Stu...              12.0000   
80389                                   Home Equity Loan              16.0000   
93445                                      Not Specified              13.0000   
22214  Student Loan, Mortgage Loan, and Debt Consolid...              21.0000   
60791  Auto Loan, Home Equity Loan, Auto Loan, Home E...              15.0000   

       Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
83285                 12.0000                8.0200                0.0000   
80389                 11.0000                2.9900                7.0000   
93445                  1.0000                4.5800                2.0000   
22214                 19.0000                8.6800                7.0000   
60791                 19.0000                3.4200                6.0000   

      Credit_Mix  Outstanding_Debt  Credit_Utilization_Ratio  \
83285   Standard          809.3200                   30.4469   
80389   Standard         1110.6300                   25.7928   
93445       Good          757.2500                   29.3420   
22214   Standard          240.2700                   33.9438   
60791        Bad         2708.5500                   28.6870   

       Credit_History_Age Payment_of_Min_Amount  Total_EMI_per_month  \
83285            273.0000                    No             121.4190   
80389            384.0000                    No              15.5226   
93445            246.0000                    No              16.8127   
22214            250.0000                    NM              96.5553   
60791            152.0000                   Yes              76.5442   

       Amount_invested_monthly                 Payment_Behaviour  \
83285                  39.8068   High_spent_Large_value_payments   
80389                  24.6217   High_spent_Small_value_payments   
93445                  32.1225  High_spent_Medium_value_payments   
22214                  49.4663    Low_spent_Large_value_payments   
60791                  58.9482   High_spent_Small_value_payments   

       Monthly_Balance Credit_Score  
83285         537.6845     Standard  
80389         322.4623         Good  
93445         490.5089     Standard  
22214         483.5177     Standard  
60791         258.8403     Standard

In [364]:
display(HTML(f"""<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> We can see the statistical characteristics of numerical features such as count, central tendency, standardeviation, minimum, maximum, and percentile value of each numerical features (columns). It will be helpful for us to understand the data distribution, stuctures, etc."""))
df.describe(include = "number")

ID  Customer_ID       Month         Age            SSN  \
count 100000.0000  100000.0000 100000.0000 100000.0000    100000.0000   
mean   80631.5000   25982.6666      4.5000     33.3163 500461680.2624   
std    43301.4866   14340.5431      2.2913     10.7648 290826734.3862   
min     5634.0000    1006.0000      1.0000     14.0000     81349.0000   
25%    43132.7500   13664.5000      2.7500     24.0000 245168577.2500   
50%    80631.5000   25777.0000      4.5000     33.0000 500688611.5000   
75%   118130.2500   38385.0000      6.2500     42.0000 756002666.2500   
max   155629.0000   50999.0000      8.0000     56.0000 999993421.0000   

       Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
count    100000.0000            100000.0000        100000.0000   
mean      50505.1234              4197.2708             5.3688   
std       38299.4221              3186.4325             2.5933   
min        7005.9300               303.6454             0.0000   
25%       19342.9725              1626.5942             3.0000   
50%       36999.7050              3095.9050             5.0000   
75%       71683.4700              5957.7150             7.0000   
max      179987.2800             15204.6333            11.0000   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  Delay_from_due_date  \
count      100000.0000    100000.0000  100000.0000          100000.0000   
mean            5.5336        14.5321       3.5329              21.0814   
std             2.0671         8.7413       2.4464              14.8046   
min             0.0000         1.0000       0.0000               0.0000   
25%             4.0000         7.0000       2.0000              10.0000   
50%             5.0000        13.0000       3.0000              18.0000   
75%             7.0000        20.0000       5.0000              28.0000   
max            11.0000        34.0000       9.0000              62.0000   

       Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
count             100000.0000           100000.0000           100000.0000   
mean                  13.3131               10.4703                5.7983   
std                    6.2372                6.6095                3.8678   
min                    0.0000                0.5000                0.0000   
25%                    9.0000                5.3800                3.0000   
50%                   14.0000                9.4000                5.0000   
75%                   18.0000               14.8500                8.0000   
max                   25.0000               29.9800               17.0000   

       Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
count       100000.0000               100000.0000         100000.0000   
mean          1426.2204                   32.2852            221.2205   
std           1155.1290                    5.1169             99.6807   
min              0.2300                   20.0000              1.0000   
25%            566.0725                   28.0526            144.0000   
50%           1166.1550                   32.3058            219.0000   
75%           1945.9625                   36.4967            302.0000   
max           4998.0700                   50.0000            404.0000   

       Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  
count          100000.0000              100000.0000      100000.0000  
mean              107.6992                  55.1013         392.6976  
std               132.2671                  39.0069         201.6527  
min                 0.0000                   0.0000           0.0078  
25%                29.2689                  27.9591         267.6160  
50%                66.4623                  45.1565         333.8654  
75%               147.3926                  71.2958         463.2157  
max              1779.1033                 434.1911        1183.9307

In [365]:
display(HTML(f"""<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> We can see the statistical characteristics of categorical features such as count, central tendency,unique values, frequency of each categorical features (columns). It will be helpful for us to understand the data distribution, stuctures, etc."""))
df.describe(exclude = "number")

Name Occupation Type_of_Loan Credit_Mix Payment_of_Min_Amount  \
count     100000     100000       100000     100000                100000   
unique     10128         15         6261          3                     3   
top     Jessicad     Lawyer      No Data   Standard                   Yes   
freq          48       7096        11408      45848                 52326   

                     Payment_Behaviour Credit_Score  
count                           100000       100000  
unique                               6            3  
top     Low_spent_Small_value_payments     Standard  
freq                             28616        53174

### <span style="color:Khaki;">Feature Details:

<span style="color:Tomato;font-size: 100%"> <b> 1. Customer_ID - </b></span> Generally not used in Analysis and modeling, but can be useful for data merging and identification.</p>
<span style="color:Tomato;font-size: 100%"> <b> 2. Month - </b></span> The month when the data was recorded. We can use for Time-based analysis and trend identification.</p>
<span style="color:Tomato;font-size: 100%"> <b> 3. Name - </b></span> Generally not used in modeling, but can be useful for data validation and integrity checks.</p>
<span style="color:Tomato;font-size: 100%"> <b> 5. Age - </b></span> Age of the customer. Age groups can be created for better analysis.</p>
<span style="color:Tomato;font-size: 100%"> <b> 6. SSN - </b></span> Social Security Number (SSN) is a tax id. Sensitive information. Generally, it’s used for identity verification but not in modeling.</p>
<span style="color:Tomato;font-size: 100%"> <b> 7. Occupation - </b></span> Customer's occupation. Categorical feature that can be used to understand income patterns and loan behaviors.</p>
<span style="color:Tomato;font-size: 100%"> <b> 8. Annual_Income - </b></span> Yearly income of the customer. Important feature for modeling creditworthiness and affordability.</p>
<span style="color:Tomato;font-size: 100%"> <b> 9. Monthly_Inhand_Salary - </b></span> Salary received monthly. Can be used to validate the Annual_Income and for affordability analysis.</p>
<span style="color:Tomato;font-size: 100%"> <b> 10. Num_Bank_Accounts - </b></span> Number of bank accounts held by the customer. Feature to analyze financial behavior and stability.</p>
<span style="color:Tomato;font-size: 100%"> <b> 10. Num_Credit_Card - </b></span> Number of credit cards owned. Indicator of credit behavior and potential risk.</p>
<span style="color:Tomato;font-size: 100%"> <b> 11. Interest_Rate - </b></span> Interest rate on loans or credit. Feature for modeling cost of borrowing and repayment behavior.</p>
<span style="color:Tomato;font-size: 100%"> <b> 12. Num_of_Loan - </b></span> Number of loans taken. Risk assessment feature. Higher number might indicate higher financial stress.</p>
<span style="color:Tomato;font-size: 100%"> <b> 13. Type_of_Loan - </b></span> The type of loan(s) the customer has. Categorical feature to distinguish between different borrowing behaviors.</p>
<span style="color:Tomato;font-size: 100%"> <b> 14. Delay_from_due_date - </b></span> Number of days delayed from the due date. Key feature for credit risk modeling.</p>
<span style="color:Tomato;font-size: 100%"> <b> 15. Changed_Credit_Limit - </b></span> Amount by which the credit limit was changed. Feature to analyze changes in creditworthiness.</p>
<span style="color:Tomato;font-size: 100%"> <b> 16. Num_of_Delayed_Payment - </b></span> Number of delayed payments. Direct indicator of payment behavior and credit risk.</p>

<span style="color:Tomato;font-size: 100%"> <b> 17. Num_Credit_Inquiries - </b></span> Number of credit inquiries made. High number may indicate credit-seeking behavior which is a risk factor.</p>

<span style="color:Tomato;font-size: 100%"> <b> 18. Credit_Mix - </b></span> The diversity of credit types (e.g., credit cards, loans). Categorical feature to assess the balance of credit types.

<span style="color:Tomato;font-size: 100%"> <b> 19. Outstanding_Debt </b></span> Total outstanding debt. Key feature for risk assessment and financial health.

<span style="color:Tomato;font-size: 100%"> <b> 20. Credit_Utilization_Ratio - </b></span> Ratio of credit used to credit available. Important feature for credit risk modeling.

<span style="color:Tomato;font-size: 100%"> <b> 21. Credit_History_Age - </b></span> Age of the credit history. Older credit history usually indicates better credit risk.

<span style="color:Tomato;font-size: 100%"> <b> 22. Payment_of_Min_Amount - </b></span> Whether only the minimum amount is paid. Binary feature indicating financial behavior.

<span style="color:Tomato;font-size: 100%"> <b> 23. Total_EMI_per_month - </b></span> Total monthly installment amount. Indicator of monthly financial burden.

<span style="color:Tomato;font-size: 100%"> <b> 24. Amount_invested_monthly - </b></span> Amount invested monthly. Financial health indicator, savings behavior.

<span style="color:Tomato;font-size: 100%"> <b> 25. Payment_Behaviour - </b></span> General payment behavior. Categorical feature for credit risk assessment.

<span style="color:Tomato;font-size: 100%"> <b> 26. Monthly_Balance - </b></span> Monthly balance left after expenses. Indicator of financial stability.

<span style="color:Tomato;font-size: 100%"> <b> 27. Credit_Score - </b></span> Credit score of the customer. Target variable for modeling creditworthiness.


### <span style="color:Khaki;"> Checking data Type:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output below, we can see the data type of each feature. We can reduce memory usage by maintaining or casting applicable data types to the features. This process will smooth our data wrangling efforts.

In [366]:
df.dtypes

ID                            int64
Customer_ID                   int64
Month                         int64
Name                         object
Age                         float64
SSN                         float64
Occupation                   object
Annual_Income               float64
Monthly_Inhand_Salary       float64
Num_Bank_Accounts           float64
Num_Credit_Card             float64
Interest_Rate               float64
Num_of_Loan                 float64
Type_of_Loan                 object
Delay_from_due_date         float64
Num_of_Delayed_Payment      float64
Changed_Credit_Limit        float64
Num_Credit_Inquiries        float64
Credit_Mix                   object
Outstanding_Debt            float64
Credit_Utilization_Ratio    float64
Credit_History_Age          float64
Payment_of_Min_Amount        object
Total_EMI_per_month         float64
Amount_invested_monthly     float64
Payment_Behaviour            object
Monthly_Balance             float64
Credit_Score                

In [367]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  int64  
 1   Customer_ID               100000 non-null  int64  
 2   Month                     100000 non-null  int64  
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  float64
 5   SSN                       100000 non-null  float64
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  float64
 8   Monthly_Inhand_Salary     100000 non-null  float64
 9   Num_Bank_Accounts         100000 non-null  float64
 10  Num_Credit_Card           100000 non-null  float64
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  float64
 13  Type_of_Loan              100000 non-null  ob

### <span style="color:Khaki;">Checking Null Values:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> This code checks and return the empty cells in our dataset. It is crusial to handle this null values to feef our data to Machine Learning algorithm.

In [368]:
print(df.isnull().sum().sum(),' -- ',df.isna().sum().sum())

0  --  0


### <span style="color:Khaki;">Checking unique values of the features:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> This code provides the unique datapoints that we have in the respective features and also checks for null values.

In [369]:
for col in df.columns:
    print(f"{col} - {len(df[col].unique())} - {df[col].nunique()} \n  Diff is {df[col].nunique() - len(df[col].unique())}") 

ID - 100000 - 100000 
  Diff is 0


Customer_ID - 12500 - 12500 
  Diff is 0
Month - 8 - 8 
  Diff is 0
Name - 10128 - 10128 
  Diff is 0
Age - 43 - 43 
  Diff is 0
SSN - 12500 - 12500 
  Diff is 0
Occupation - 15 - 15 
  Diff is 0
Annual_Income - 12488 - 12488 
  Diff is 0
Monthly_Inhand_Salary - 13241 - 13241 
  Diff is 0
Num_Bank_Accounts - 12 - 12 
  Diff is 0
Num_Credit_Card - 12 - 12 
  Diff is 0
Interest_Rate - 34 - 34 
  Diff is 0
Num_of_Loan - 10 - 10 
  Diff is 0
Type_of_Loan - 6261 - 6261 
  Diff is 0
Delay_from_due_date - 63 - 63 
  Diff is 0
Num_of_Delayed_Payment - 26 - 26 
  Diff is 0
Changed_Credit_Limit - 3452 - 3452 
  Diff is 0
Num_Credit_Inquiries - 18 - 18 
  Diff is 0
Credit_Mix - 3 - 3 
  Diff is 0
Outstanding_Debt - 12203 - 12203 
  Diff is 0
Credit_Utilization_Ratio - 100000 - 100000 
  Diff is 0
Credit_History_Age - 404 - 404 
  Diff is 0
Payment_of_Min_Amount - 3 - 3 
  Diff is 0
Total_EMI_per_month - 11890 - 11890 
  Diff is 0
Amount_invested_monthly - 12261 - 12261 
  Diff is 0
Payment_Behav

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the output above, we can see that we have 12500 customer details in the overall 100000 data points.<br>
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Customer name count 10128 confirms that the customers may have multiple accounts.<br>
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Credit_Utilization_Ratio and ID are unique value. It may not be useful.

### <span style="color:Khaki;">Checking values lesser than or equal to 0:

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Sometimes the data may contain negative values or just updated with 0 due to data entry error. This code will show such datapoints.

In [370]:
for col in df.select_dtypes(include=['number']).columns:
    print(f"{col} --  {(df[col] <= 0).sum()}")

ID --  0
Customer_ID --  0
Month --  0
Age --  0
SSN --  0
Annual_Income --  0
Monthly_Inhand_Salary --  0
Num_Bank_Accounts --  4417
Num_Credit_Card --  14
Interest_Rate --  0
Num_of_Loan --  11408
Delay_from_due_date --  1385
Num_of_Delayed_Payment --  2081
Changed_Credit_Limit --  0
Num_Credit_Inquiries --  7190
Outstanding_Debt --  0
Credit_Utilization_Ratio --  0
Credit_History_Age --  0
Total_EMI_per_month --  10985
Amount_invested_monthly --  1920
Monthly_Balance --  0


<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From above output, few of the features may have negative values or zero. We must check and address any errors.

### <span style="color:Khaki;">Initial Observations:


1. ID feature may not be useful as these are unique datapoints. So we can drop these features.
2. The Type_of_Loan column has array values. We may get additional information if we extract number of loans.
3. From the dataset statistics, minimum age is 11. Possibly it is an error. We must explore this feature.
5. The interest rate ranges from 1 to 34, which is broad. Validate the upper limit to ensure it's within expected bounds.
1. Monthly salary calculated from annual income doesn't match the reported monthly in-hand salary. For example, the mean annual income divided by 12 doesn't equal the mean monthly in-hand salary, which may indicate bonuses or other factors.
1. The minimum of $0.23 seems unusually low for outstanding debt.
1. The minimum of $303.65 per month is extremely low and might indicate part-time work or errors.
1. The maximum values seem high (11), but they could be realistic for some customers.
1. The maximum number of loans is 9, which is high. We must explore the data more to verify.
1. The maximum number of delayed payments is 25. We must verify this this exploring the data.
1. The maximum EMI per month seems high (1779.1033). We must verify.
1. The maximum value (434.1911) is significantly higher than the 75th percentile (71.2958). There might be outliers.
6. We could potentially create a new features like below:
8. Need to verify the clasess in the categorical features to ensure consistency.
8. Few of the features may have negative values or zero. We must check and address any errors.
9. Need to convert data type for efficient storage.
8. Need to verify the clasess in the categorical features to ensure consistency.


    - <span style="color: DodgerBlue;font-family:calibri;">**Debt_to_Income_Ratio -**</span>  Annual_Income / Num_of_Loan. This feature indicates the financial health of a customer by comparing their Outstanding_Debt to their Annual_Income.
    - <span style="color: DodgerBlue;font-family:calibri;">**Income_to_Loan_Ratio**</span> - Outstanding_Debt / Annual_Income. This feature can be created by dividing the Annual_Income by the Num_of_Loan to understand how many loans a customer has relative to their income. 
    - <span style="color: DodgerBlue;font-family:calibri;">**Monthly_Debt_Payment**</span> - Outstanding_Debt / Num_of_Loan.  Estimate the monthly debt payment by dividing Outstanding_Debt by Num_of_Loan.
    - <span style="color: DodgerBlue;font-family:calibri;">**Credit_Utilization_per_Account**</span> - Credit_Utilization_Ratio / Num_Bank_Accounts. This feature divides the Credit_Utilization_Ratio by the Num_Bank_Accounts to get an average credit utilization per account.   
    - <span style="color: DodgerBlue;font-family:calibri;">**Credit_Inquiry_per_Loan**</span> - Num_Credit_Inquiries / Num_of_Loan. This feature shows how many credit inquiries were made per loan.
    - <span style="color: DodgerBlue;font-family:calibri;">**Age_Group**</span> - Categorical age groups/bins.
    - <span style="color: DodgerBlue;font-family:calibri;">**Loan_to_Income_Ratio**</span> - Total_EMI_per_month / Monthly_Inhand_Salary. Compare the total EMI per month to the monthly in-hand salary.
    - <span style="color: DodgerBlue;font-family:calibri;">**Delayed_Payment_Ratio**</span> - Num_of_Delayed_Payment / Num_of_Loan. Compare the number of delayed payments to the total number of loans.
    - <span style="color: DodgerBlue;font-family:calibri;">**Investment_to_Income_Ratio**</span> - Amount_invested_monthly / Monthly_Inhand_Salary. Calculate the ratio of the amount invested monthly to the monthly in-hand salary.
    - <span style="color: DodgerBlue;font-family:calibri;">**Credit_Limit_Change_Rate**</span> - Changed_Credit_Limit/ Credit_History_Age. Using month, calculate the rate of change in credit limit over time.
    - <span style="color: DodgerBlue;font-family:calibri;">**Payment_Behavior_Category**</span> - Good or Bad. Convert Payment_Behaviour to a numerical or categorical feature indicating good, bad, or neutral behavior.
    - <span style="color: DodgerBlue;font-family:calibri;">**Total_Credit_Exposure**</span> - Outstanding_Debt / Credit_Utilization_Ratio. Calculate the total credit exposure by summing up outstanding debt and credit utilization ratio.
    - <span style="color: DodgerBlue;font-family:calibri;">**Credit_Utilization_to_Income_Ratio:**</span> - This could show how much of their income is tied up in credit usage.
    - <span style="color: DodgerBlue;font-family:calibri;">**Savings_rate**</span> - Amount_invested_monthly / Monthly_Inhand_Salary. This indicates the proportion of income being saved or invested.
    - <span style="color: DodgerBlue;font-family:calibri;">**Reliable_payment**</span> - (Num_of_Delayed_Payment / Credit_History_Age) * 100. This could indicate how often payments are delayed over time.
    - <span style="color: DodgerBlue;font-family:calibri;">**Disposable_income**</span> - Monthly_Inhand_Salary - Total_EMI_per_month - Amount_invested_monthly. This shows how much money is left after obligations and investments.
    - <span style="color: DodgerBlue;font-family:calibri;">**income_stability**</span> - Monthly_Inhand_Salary / (Annual_Income / 12). If this is not close to 1, it might indicate variable income.
    - <span style="color: DodgerBlue;font-family:calibri;">**Financial_stress**</span> - (Total_EMI_per_month + Outstanding_Debt) / Monthly_Inhand_Salary. This could indicate how stretched a person's finances are.investments.
    - <span style="color: DodgerBlue;font-family:calibri;">**Debt Payoff Time:**</span> - Outstanding_Debt / Total_EMI_per_month. This estimates how many months it would take to pay off debt at current EMI.

### <span style="color:Khaki;">Removing unecessary columns and updating column Names:

In [371]:
df.columns # to print all columns names in our data

Index(['ID', 'Customer_ID', 'Month', 'Name', 'Age', 'SSN', 'Occupation',
       'Annual_Income', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts',
       'Num_Credit_Card', 'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan',
       'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

In [372]:
# df.drop(columns=['ID','SSN'],axis=1,inplace=True) # removing ID and SSN from the dataset
df = df.rename(columns=str.lower) # changing column names to lower case for easy access
df.columns # printing column names to check 

Index(['id', 'customer_id', 'month', 'name', 'age', 'ssn', 'occupation',
       'annual_income', 'monthly_inhand_salary', 'num_bank_accounts',
       'num_credit_card', 'interest_rate', 'num_of_loan', 'type_of_loan',
       'delay_from_due_date', 'num_of_delayed_payment', 'changed_credit_limit',
       'num_credit_inquiries', 'credit_mix', 'outstanding_debt',
       'credit_utilization_ratio', 'credit_history_age',
       'payment_of_min_amount', 'total_emi_per_month',
       'amount_invested_monthly', 'payment_behaviour', 'monthly_balance',
       'credit_score'],
      dtype='object')

### <span style="color:Khaki;">Updating Data types:
<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> Casting appropriate data types reduces dataset memeory and allow us to perform data cleaning operation smoothly.

In [373]:
df.describe() # checking statistical information to cast appropriate datatype.

id  customer_id       month         age            ssn  \
count 100000.0000  100000.0000 100000.0000 100000.0000    100000.0000   
mean   80631.5000   25982.6666      4.5000     33.3163 500461680.2624   
std    43301.4866   14340.5431      2.2913     10.7648 290826734.3862   
min     5634.0000    1006.0000      1.0000     14.0000     81349.0000   
25%    43132.7500   13664.5000      2.7500     24.0000 245168577.2500   
50%    80631.5000   25777.0000      4.5000     33.0000 500688611.5000   
75%   118130.2500   38385.0000      6.2500     42.0000 756002666.2500   
max   155629.0000   50999.0000      8.0000     56.0000 999993421.0000   

       annual_income  monthly_inhand_salary  num_bank_accounts  \
count    100000.0000            100000.0000        100000.0000   
mean      50505.1234              4197.2708             5.3688   
std       38299.4221              3186.4325             2.5933   
min        7005.9300               303.6454             0.0000   
25%       19342.9725              1626.5942             3.0000   
50%       36999.7050              3095.9050             5.0000   
75%       71683.4700              5957.7150             7.0000   
max      179987.2800             15204.6333            11.0000   

       num_credit_card  interest_rate  num_of_loan  delay_from_due_date  \
count      100000.0000    100000.0000  100000.0000          100000.0000   
mean            5.5336        14.5321       3.5329              21.0814   
std             2.0671         8.7413       2.4464              14.8046   
min             0.0000         1.0000       0.0000               0.0000   
25%             4.0000         7.0000       2.0000              10.0000   
50%             5.0000        13.0000       3.0000              18.0000   
75%             7.0000        20.0000       5.0000              28.0000   
max            11.0000        34.0000       9.0000              62.0000   

       num_of_delayed_payment  changed_credit_limit  num_credit_inquiries  \
count             100000.0000           100000.0000           100000.0000   
mean                  13.3131               10.4703                5.7983   
std                    6.2372                6.6095                3.8678   
min                    0.0000                0.5000                0.0000   
25%                    9.0000                5.3800                3.0000   
50%                   14.0000                9.4000                5.0000   
75%                   18.0000               14.8500                8.0000   
max                   25.0000               29.9800               17.0000   

       outstanding_debt  credit_utilization_ratio  credit_history_age  \
count       100000.0000               100000.0000         100000.0000   
mean          1426.2204                   32.2852            221.2205   
std           1155.1290                    5.1169             99.6807   
min              0.2300                   20.0000              1.0000   
25%            566.0725                   28.0526            144.0000   
50%           1166.1550                   32.3058            219.0000   
75%           1945.9625                   36.4967            302.0000   
max           4998.0700                   50.0000            404.0000   

       total_emi_per_month  amount_invested_monthly  monthly_balance  
count          100000.0000              100000.0000      100000.0000  
mean              107.6992                  55.1013         392.6976  
std               132.2671                  39.0069         201.6527  
min                 0.0000                   0.0000           0.0078  
25%                29.2689                  27.9591         267.6160  
50%                66.4623                  45.1565         333.8654  
75%               147.3926                  71.2958         463.2157  
max              1779.1033                 434.1911        1183.9307

#### <span style="color:Khaki;">Updating data type for numeric features.

In [374]:
dtype_dict = {
    'month': 'int8',
    'age': 'int8',
    'num_bank_accounts': 'int8',
    'num_credit_card': 'int8',
    'interest_rate': 'int8',
    'num_of_loan': 'int8',
    'delay_from_due_date': 'int8',
    'num_of_delayed_payment': 'int8',
    'changed_credit_limit': 'int8',
    'credit_utilization_ratio': 'int8',
    'credit_history_age': 'int32',
    'annual_income': 'float64',
    'monthly_inhand_salary': 'float64',
    'total_emi_per_month': 'float64',
    'amount_invested_monthly': 'float64',
    'monthly_balance': 'float64',
    'customer_id': 'int64'
}

df = df.astype(dtype_dict)

#### <span style="color:Khaki;">Updating data type for catagorical features.

In [375]:
for col in df.select_dtypes(exclude='number').columns:
    df[col] = df[col].astype('category')

In [376]:
df.info() # checking datatypes and memory usage after changing datatypes.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 28 columns):
 #   Column                    Non-Null Count   Dtype   
---  ------                    --------------   -----   
 0   id                        100000 non-null  int64   
 1   customer_id               100000 non-null  int64   
 2   month                     100000 non-null  int8    
 3   name                      100000 non-null  category
 4   age                       100000 non-null  int8    
 5   ssn                       100000 non-null  float64 
 6   occupation                100000 non-null  category
 7   annual_income             100000 non-null  float64 
 8   monthly_inhand_salary     100000 non-null  float64 
 9   num_bank_accounts         100000 non-null  int8    
 10  num_credit_card           100000 non-null  int8    
 11  interest_rate             100000 non-null  int8    
 12  num_of_loan               100000 non-null  int8    
 13  type_of_loan              1000

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> After casting data type we could see that the memory usage is reduced to 8.8 MB.

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> From the above output we see

### <p style="color: Khaki;">Cleaning Case and Text inconsistency and checking unique values:</p>

In [377]:
for col in df.select_dtypes(include='category'):
    if 'block' not in col:
        df[col] = df[col].str.strip().str.title()
        df[col].describe()
        print(df[col].describe(),'\n\n',df[col].value_counts().reset_index().sort_values(by=col,ascending=True),'\n','-----'*10)

count       100000
unique       10128
top       Jessicad
freq            48
Name: name, dtype: object 

                        name  count
8613              A Coopert      8
2526             A Maxwella      8
8192     A. Ananthalakshmih      8
10114    A. Ananthalakshmiq      8
4384     A. Ananthalakshmix      8
...                     ...    ...
3751             Zieminskis      8
7760             Zieminskiv      8
5525                Zinetsn      8
4248   Zoran Radosavljevicl      8
5938                 Zorans      8

[10128 rows x 2 columns] 
 --------------------------------------------------
count     100000
unique        15
top       Lawyer
freq        7096
Name: occupation, dtype: object 

        occupation  count
5      Accountant   6744
2       Architect   6824
6       Developer   6720
10         Doctor   6568
1        Engineer   6864
9    Entrepreneur   6648
11     Journalist   6536
0          Lawyer   7096
12        Manager   6432
3        Mechanic   6776
7   Media_Manager 

<span style="color: Chartreuse;"> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &#9784; </span> <span style="color: white;">From the above output, type_of_loan feature has array of values (clustered value). It will not be useful to our model as it recognizes this as unique text. So we are going to split each loan into each column to provide weightage based on the loan count. Finally, it creates columns with the all loan type and provide value to rows as 1 if someone has the loan or it will be 0 (One-Hot Encoding).

In [378]:
# Splitting type of loans based on the comma

df['type_of_loan'] = df['type_of_loan'].apply(lambda x: x if isinstance(x, list) else x.split(','))

# Removing 'And' and white spaces 
df['type_of_loan'] = df['type_of_loan'].apply(lambda x: list(set([loan.replace('And ', '').strip()  for loan in x])))


In [379]:
# One-Hot Encoding
loan_types = set(loan for loans in df['type_of_loan'] for loan in loans)
for loan_type in loan_types:
    df[loan_type] = df['type_of_loan'].apply(lambda x: 1 if loan_type in x else 0)
    
# Dropping original column and other loan type
df.drop(['type_of_loan','Not Specified','No Data'], axis=1, inplace=True)

In [380]:
df.sample(2)

id  customer_id  month      name  age            ssn occupation  \
2327    9121        27049      8  Blamontg   19 399143062.0000     Writer   
40005  65639        16830      6  Selyukha   35 700015042.0000   Musician   

       annual_income  monthly_inhand_salary  num_bank_accounts  \
2327      97640.4000              8034.7000                  5   
40005     72583.3200              6093.6100                  7   

       num_credit_card  interest_rate  num_of_loan  delay_from_due_date  \
2327                 7              6            4                   19   
40005                3             10            2                   27   

       num_of_delayed_payment  changed_credit_limit  num_credit_inquiries  \
2327                       14                     9                4.0000   
40005                      13                     7                4.0000   

      credit_mix  outstanding_debt  credit_utilization_ratio  \
2327    Standard           80.5000                        25   
40005   Standard         1019.4400                        31   

       credit_history_age payment_of_min_amount  total_emi_per_month  \
2327                  137                   Yes             259.8059   
40005                 223                   Yes             104.0167   

       amount_invested_monthly                payment_behaviour  \
2327                   99.7393  Low_Spent_Medium_Value_Payments   
40005                  45.3392   Low_Spent_Small_Value_Payments   

       monthly_balance credit_score  Mortgage Loan  Home Equity Loan  \
2327          467.1766     Standard              0                 0   
40005         529.6125     Standard              0                 0   

       Credit-Builder Loan  Payday Loan  Personal Loan  \
2327                     1            0              0   
40005                    0            0              1   

       Debt Consolidation Loan  Student Loan  Auto Loan  
2327                         0             1          1  
40005                        0             0          1

<p style="color: Khaki;">

## <p Style="color: Aqua"> Exporting Data Frame

In [381]:
df.to_feather('Cleaned_data') # Exporting data frame as feather data type. It is efficient and keep our data types.